Kliff master v1 lightning

ipcamit opened the pull request:
## Summary

Added base trainer and lightning trainer. Along with new tests to test the trainer module

## Additional dependencies introduced (if any)

* torch lightning for lightning trainer in ["test"]

## TODO (if any)

Add KIM Trainer next

## Checklist

Before a pull request can be merged, the following items must be checked:

* [x] Make sure your code is properly formatted. [isort](https://pycqa.github.io/isort/) and [black](https://black.readthedocs.io/en/stable/getting_started.html) are used for this purpose. The simplest way is to use [pre-commit](https://pre-commit.com). See instructions [here](https://github.com/openkim/kliff/blob/main/docs/source/contributing_guide.md#code-style).
* [x] Doc strings have been added in the [Google docstring format](https://www.sphinx-doc.org/en/master/usage/extensions/example_google.html) on your code.
* [x] Type annotations are **highly** encouraged. Run [mypy](http://mypy-lang.org) to
  type check your code.
* [x] Tests have been added for any new functionality or bug fixes.
* [x] All linting and tests pass.

Note that the CI system will run all the above checks. But it will be much more
efficient if you already fix most errors prior to submitting the PR.

mjwen commented on kliff/trainer/lightning_trainer.py:
```diff
@@ -0,0 +1,514 @@
+            grad_outputs=torch.ones_like(predicted_energy),
+        )
+        predicted_forces = scatter_add(predicted_forces, batch["images"], dim=0)
+        return predicted_energy, -predicted_forces
```
Lets move the negative sign to one of the above line where `predicted_forces` is computed?

mjwen commented on kliff/trainer/lightning_trainer.py:
```diff
@@ -0,0 +1,514 @@
+            [predicted_energy],
+            batch["coords"],
+            create_graph=True,  # TODO: grad against arbitrary param name
+            retain_graph=True,
```
`retain_graph` should be `False`. See docs here: https://pytorch.org/docs/stable/generated/torch.autograd.grad.html#torch.autograd.grad

mjwen commented on kliff/trainer/lightning_trainer.py:
```diff
@@ -0,0 +1,514 @@
+        model_inputs = {k: batch[k] for k in self.input_args}
+        predicted_energy = self.model(**model_inputs)
+        (predicted_forces,) = torch.autograd.grad(
+            [predicted_energy],
```
No need to use `[predicted_energy]`; instead `predicted_energy.sum()` can be used. With this, we don't need to provide `grad_outputs` either. So, this line becomes. 

```
(predicted_forces,) = torch.autograd.grad(
            predicted_energy.sum(),
            batch["coords"],
            create_graph=True, 
)       
```

ipcamit replied:
Thats a neat trick, did not think of it. thanks!

mjwen commented on kliff/trainer/lightning_trainer.py:
```diff
@@ -0,0 +1,514 @@
+            + forces_weight * torch.mean(per_atom_force_loss) / 3
+        )  # divide by 3 to get correct MSE
+
+        self.log(
```
Besides a sum of the loss, people would typically be more interested in separate losses on energy and forces. 

Also, people can also be interested in metrics like MAE other than MSE.

Most importantly, it is not interesting to know losses/metrics of a validation step for a batch of data. The losses/metrics at each epoch for all data matter. This requires aggregating the data between steps. I'd suggest using torchmetrics, which automatically does it.   

So, we need to report separate losses, and allow other metrics, e.g. using the [trochmetrics](https://lightning.ai/docs/torchmetrics/stable/) package.

ipcamit replied:
Torch metrics looks great. I will add support for it. 
Can you explain this a bit:
> Most importantly, it is not interesting to know losses/metrics of a validation step for a batch of data. The losses/metrics at each epoch for all data matter.

mjwen replied:
Currently, the validation loss/metrics is evaluated and reported for each mini batch of data, which is only part of all the validation data. But people generally would be interested in the loss/metrics for all validation/test data, not each mini batch.

ipcamit replied:
I dont think so. In logging when we give `on_epoch=True` and `on_step=False`, then I believe the logger will call `after_batch_end` function to simply accumulate the loss and only log it in `after_epoch_end` function. In the log files too I could see that validation losses are logged per epoch, and not per batch. This is similar to what I am doing in the loss traj callback, where `on_validation_batch_end` just gathers the result and `on_validation_epoch_end` writes them to a file. I can recheck the API but I am quite certain that is the case.

mjwen commented:
@ipcamit Nothing major, but a few clarifying questions and minor tweaks.

ipcamit commented on kliff/dataset/dataset.py:
```diff
@@ -969,17 +1027,300 @@ def __len__(self) -> int:
+        )
+
+    @staticmethod
+    def get_manifest_checksum(
```
1. Yes the recursive keys are sorted,
```
In [7]: json.dumps({"a":2, "b":{"b":3,"a":1,"c":{"b":1,"d":10,"a":22}}},sort_keys=True)
Out[7]: '{"a": 2, "b": {"a": 1, "b": 3, "c": {"a": 22, "b": 1, "d": 10}}}'
```
2. I dont fully understand this point, `dataset_manifest` will have default values always. So missing keys will have predictable values. The checksum will only change when users have provided some custom keys.

mjwen replied:
1. Great! 
2. This boils down to whether `dataset_manifest` consists of all allowed settings or the settings provided by a user. If the former, there is no problem. But if the latter, there could be two different checksums corresponding to the same training, if a user chooses not to provide some values but uses the default. I guess it is the former, so no problem then.

ipcamit commented on kliff/models/kim.py:
Yes. Lightning trainer will be the most used one I think, so better to get it right, then make others work. KIM physics based trainer is working. Will add it in next PR.

ipcamit commented on setup.py:
```diff
@@ -126,6 +126,8 @@ def get_readme():
             "libdescriptor",
             "torch_geometric",
+            "pytorch_lightning",
+            "dill",
```
Done. Added torch as additional install option.

ipcamit commented on kliff/transforms/configuration_transforms/graphs/generate_graph.py:
```diff
@@ -14,19 +14,26 @@
 if torch_geometric_available():
     from torch_geometric.data import Data
+else:
+    Data = object
```
`Data` is used as a parent class in `PyGGraph`. Otherwise I get error in importing `transforms` module.

mjwen replied:
Sounds good.

ipcamit commented on kliff/trainer/torch_trainer_utils/lightning_checkpoints.py:
```diff
@@ -0,0 +1,92 @@
+from kliff.dataset import Dataset
+
+
+class SaveModelCallback(pl.Callback):
```
Yes. Perhaps I can add it in next round of improvements. As it will include moving and deleting files on storage so need to be more cautious. Leaving this comment unresolved for a TODO later.

ipcamit commented on kliff/trainer/torch_trainer_utils/lightning_checkpoints.py:
```diff
@@ -0,0 +1,92 @@
+        trainer.save_checkpoint(os.path.join(self.ckpt_dir, "trainer_checkpoint.ckpt"))
+
+
+class LossTrajectoryCallback(pl.Callback):
```
None of the existing loggers quite solve this problem. But I think it is a good idea to write own logger for this using lightning logger class. Will write one and try.

mjwen replied:
What functionality do you want to have but is not provided by the default loggers?

ipcamit replied:
This feature needs to be more fleshed out, which I would do after discussion with Josh (as he is the one who needs it). His requirement is to track atomwise losses in _both training and validation_ dataset and use it as a measure in some form of active learning where most pathological configurations are given different weightage (that is how I understood it). 
The biggest limitations I was thinking of at the time was 
i. binary format, for longer training epochs, or larger configurations, we might switch to pkl, or npy binary format later on, as otherwise the file sizes balloon up a lot.
ii. logging the atomwise losses, along with indexes. At present I keep the validation dataset unshuffled and save the losses, but ideally I would like to save it basex on `idx` from the trainer. I could not figure out a way to use existing logger like CSVLogger to log this information.

I tried writing a custom logger yesterday, but now I feel that using a CSV or custom logger might result in lot of twisted code, and callback might be cleanest solution.

I think overall problem is that it is somewhat underspecified problem that I am keeping here for making sure it exists in the API what we can improve upon as needed.

ipcamit commented on kliff/trainer/torch_trainer_utils/dataloaders.py:
```diff
@@ -0,0 +1,173 @@
+from kliff.transforms.configuration_transforms.graphs import PyGGraph
+
+
+class NeighborListDataset(TorchDataset):
```
This is more of a future-proofing for models where we might not know the exact requirements beforehand. So this provides the raw processed data that users can use in any way they see fit.

ipcamit commented on kliff/trainer/base_trainer.py:
```diff
@@ -0,0 +1,768 @@
+        directory for the current run. It will be the hash of the configuration dictionary
+        string.
+        """
+        config = self.config_to_dict()
```
Yes. For now no, we are not comparing them. I am not sure if I will ever use this feature for comparing. but it is just a way to assign unique id to a run configuration. Idea is to use this for archiving the trainer script and hash as a mean of provenance stating that the submitted kim model used this settings

ipcamit commented on kliff/trainer/base_trainer.py:
```diff
@@ -0,0 +1,768 @@
+        """
+        np.random.seed(self.workspace["seed"])
+        random.seed(self.workspace["seed"])
+        # torch.manual_seed(self.seed) # enable torch seed in respective children
```
Yes. it was a TODO. Will add it

ipcamit commented on kliff/trainer/base_trainer.py:
```diff
@@ -0,0 +1,768 @@
+            self.dataset_manifest["path"],
+            num_chunks=self.optimizer_manifest["num_workers"],
+            energy_key=self.dataset_manifest.get("keys", {}).get("energy", "energy"),
+            forces_key=self.dataset_manifest.get("keys", {}).get("forces", "forces"),
```
Yes. I will need some more time for it. Perhaps v1.1 or 1.0.1? I had discussion with Ellad over best way forward. Stress for ML models is bit tricky to get right in KIM API framework, infact stresses are tricky for ML models in general. He gave some suggestions that I will like to try after the design of trainers is finalized.

mjwen replied:
Sounds good

ipcamit commented on kliff/trainer/base_trainer.py:
```diff
@@ -0,0 +1,768 @@
+        """
+        raise TrainerError("setup_optimizer not implemented.")
+
+    def setup_dataset_split(self):
```
OK. Will add in next update when I merge KIM trainer. Leaving this unresolved for now.

ipcamit commented on kliff/trainer/base_trainer.py:
```diff
@@ -0,0 +1,768 @@
+            energy_key=self.dataset_manifest.get("keys", {}).get("energy", "energy"),
+            forces_key=self.dataset_manifest.get("keys", {}).get("forces", "forces"),
+        )
+        self.dataset = deepcopy(dataset_list[0])
```
Because we want to merge the multiprocessed list of configurations into a single dataset. So I copied the fist one and extended it with remaining chunks. The list contains several datasets (no configurations) So I cant directly use it.

ipcamit commented on kliff/trainer/lightning_trainer.py:
```diff
@@ -0,0 +1,514 @@
+# Mostly to be used by GNNs for now
+# check torch version, if <= 1.13, use torch_geometric.data.lightning_module
```
TorchML model driver has not been tested with torch > 2.0 yet. I think I will update it when I test and update the TorchML model driver.

ipcamit commented on kliff/trainer/lightning_trainer.py:
```diff
@@ -0,0 +1,514 @@
+from copy import deepcopy
+from typing import Any, Dict, List, Tuple, Union
+
+import pytorch_lightning as pl
```
There was a bug in pytorch gemoetric `LightningDataset`, where it still uses the older `pytorch_lightning` namespace. If I use ligthning it gives assertion error. I think we should move to `lightning` when torch geometric does so. [See](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/data/lightning/datamodule.html#LightningDataset)

ipcamit commented on kliff/trainer/lightning_trainer.py:
```diff
@@ -0,0 +1,514 @@
+
+from .base_trainer import Trainer, TrainerError
+
+if importlib.metadata.version("torch") <= "1.13":
```
Perhaps after testing the TorchML model driver on latest torch version?

ipcamit commented on kliff/trainer/lightning_trainer.py:
```diff
@@ -0,0 +1,514 @@
+    """
+    Wrapper class for Pytorch Lightning Module. This class is used to wrap the model and
+    the training loop in a Pytorch Lightning Module. It returns the energy, and forces
+    predicted by the model. Stress computations need partial-derivatives of the
```
It is the issue with the stress, according to Ellad for periodic systems we need only the forces of the core cell, and ignore the contributions from the ghost atoms, or periodic images, then compute $f \otimes r$. He was calling this derivative of energy with only the core cell particles as partial derivatives, and summed derivatives with ghost atoms as derivatives. So I used the same terminology

ipcamit commented on kliff/trainer/lightning_trainer.py:
```diff
@@ -0,0 +1,514 @@
+        optimizer_name: Name of the optimizer to use. Default is "Adam"
+        lr: Learning rate for the optimizer. Default is 0.001
+        energy_weight: Weight for the energy loss. Default is 1.0
+        forces_weight: Weight for the forces loss. Default is 1.0
```
TODO: once the stress issue is fixed.

ipcamit commented on kliff/trainer/torch_trainer_utils/dataloaders.py:
```diff
@@ -0,0 +1,173 @@
+            dict: collated configuration, properties and contribution index tensors.
+        """
+        # get fingerprint and consistent properties
+        config_0, property_dict_0 = batch[0]
```
No, fingerprint is the complete representation of the system, including neighbors etc. This was needed to evaluate batched gradients on descriptor. It will become more clear when I integrate the DNN trainer. Leaving this open for now.

mjwen replied:
OK... then the name `fingerprint` is just a bit misleading. We might want to find another name for it

ipcamit replied:
How about `representation`?

ipcamit commented on kliff/trainer/torch_trainer_utils/dataloaders.py:
```diff
@@ -0,0 +1,173 @@
+        for prop in self.consistent_properties:
+            property_dict_0[prop] = torch.as_tensor(property_dict_0[prop])
+
+        for i in range(1, batch_len):
```
I will change this once I commit the DNN trainer (torch trainer) which uses this dataloader. Leaving it open for now.

ipcamit commented:
Addressed most comments. Need some more time and meeting with Josh for Loss trajectory concretization. Will address the descriptor dataloader issues with torch trainer which uses the descriptor module, so it will be easier to see the design choices.
